main.cpp

```diff
@@ -38,8 +39,271 @@
+            }
+            break;
+        }
+        case 1: { // Mutate data
+            if (data_maxlen > 0) {
+                size_t original_data_size = data.size();
+                // Make sure we have a buffer that's large enough for mutation
+                data.resize(std::max(data.size(), data_maxlen));
+
+                size_t new_len = LLVMFuzzerMutate(data.data(),
+                                                 original_data_size,
+                                                 data_maxlen);
+
+                new_len = std::max(new_len, size_t{1});
+                data.resize(new_len);
+
+                for (auto& val : data) {
+                    val &= 0x1F;
+                }
+            }
+            break;
+        }
+    }
+
+    if (hrp.empty() || hrp.size() > 83) {
+        return initial_input(fuzz_data, size, max_size);
+    } else if (!hrp.empty()) {
+        for (const char& c : hrp) {
+            if (c >= 'A' && c <= 'Z') {
+                return initial_input(fuzz_data, size, max_size);
+            }
+        }
+    }
+
+    if (data.empty() || data.size() > 65) {
+        return initial_input(fuzz_data, size, max_size);
+    }
+
+    std::string output = bech32::Encode(bech32::Encoding::BECH32, hrp, data);
+
+    if (output.empty() || output.length() > max_size) {
+        return initial_input(fuzz_data, size, max_size);
+    }
+
+    std::memcpy(fuzz_data, output.data(), output.length());
+    return output.length();
+}
+
+static size_t insert_part(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                        size_t in2_size, uint8_t *out, size_t max_out_size)
+{
+    if (in1_size >= max_out_size || in1_size == 0 || in2_size == 0)
+        return 0;
+
+    size_t max_insert_size = std::min(max_out_size - in1_size, in2_size);
+    size_t insert_begin = std::rand() % in1_size;
+    size_t insert_size = (std::rand() % max_insert_size) + 1;
+    size_t in2_begin = std::rand() % (in2_size - insert_size + 1);
+
+    std::memcpy(out, in1, insert_begin);
+    std::memcpy(out + insert_begin, in2 + in2_begin, insert_size);
+    std::memcpy(out + insert_begin + insert_size, in1 + insert_begin,
+               in1_size - insert_begin);
+
+    return in1_size + insert_size;
+}
+
+static size_t overwrite_part(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
+                           size_t in2_size, uint8_t *out, size_t max_out_size)
+{
+    if (in1_size > max_out_size || in1_size == 0)
+        return 0;
+
+    size_t overwrite_begin = std::rand() % in1_size;
+    size_t overwrite_size = (std::rand() % (in1_size - overwrite_begin)) + 1;
+    overwrite_size = std::min(overwrite_size, in2_size);
+    size_t in2_begin = std::rand() % (in2_size - overwrite_size + 1);
+
+    std::memcpy(out, in1, in1_size);
+    std::memcpy(out + overwrite_begin, in2 + in2_begin, overwrite_size);
+
+    return in1_size;
+}
+
+size_t cross_over(const uint8_t *in1, size_t in1_size, const uint8_t *in2,
```

Comment:
This function can be `static`.  Or better yet, all these helpers can go in an unnamed namespace.

Reply:
Just made it `static` but I intend to move it to the `bolt11mutator` module.